### ISSUE TITLE
Cabal support

### ISSUE THREAD

**Issue by cblgh** [RangerMauve/hyper-presence#1]
just saw this! i don't know what yr thinking with the module, but if it could be built in such a way that we could easily integrate it into cabal, that would be amazing!

here would be the relevant parts of where it is used atm / where it would be integrated
* https://github.com/cabal-club/cabal-client/blob/master/src/cabal-details.js#L724-L742
* https://github.com/cabal-club/cabal-core/blob/master/swarm.js#L25-L46
* https://github.com/cabal-club/cabal-core/blob/master/index.js#L229-L235

feel free to close this if it's out of scope, but i think it should be possible to do in a fairly unobtrusive way :)

**Comment by RangerMauve** [comment]
Ooo, that'd be cool. I was thinking of making the interface looks something like

```js
const p = await Presence.start(somehypercore, {
  id: Buffer.from('A', 32)
})
p.setData({something: 'whatever'})
p.on('peer-found', (peer) => peer.data)
p.on('peer-removed', (peer) => peer.id)
for(let peer of p.peers) {
  console.log(peer.id, peer.data, 'connected to', peer.connectedTo)
}
```

Under the hood it'll use flooding broadcasts to notify peers when somebody gets a new connection or loses a connection. Each peer will have a local graph tracking the state of all the peers and that's how it'll find peers that have gone offline / see who's connected to who.

The communication will be done through extension messages on a hypercore. 😁

Does that seem usable? Are there any easy changes that'd make it more usable for cabal?

**Comment by RangerMauve** [comment]
Oh, also the peer "data" is an arbitrary JSON blob. I'm going to be using it in Natakanu to list the archive URLs of people's accounts on a given computer. The goal there is to be able to render the graph of everyone that's currently online so you can have a sense of community and be able to see projects shared by other people in your group.

**Comment by cblgh** [comment]
@RangerMauve i *think* as long as it conforms to the hypercore-protocol interface, it should work for kappa-core too, which means it should work for cabal :)

what we do atm is we just use the replication stream and [call `registerExtension`](https://github.com/cabal-club/cabal-core/blob/master/swarm.js#L34-L42)

tldr yeah prolly looks good ^_^

**Comment by ghost** [comment]
I think it's better to not take a hypercore or hypercore-protocol instance for this but instead return an object that can be fed into `feed.registerExtension(name, ext)` or `proto.registerExtension(name, ext)`.

This is what I did for hypercore-query-extension: https://github.com/peermaps/hypercore-query-extension/#qextension

This will work for hypercores and hypercore-protocol instances and gives users of this module more flexibility with naming schemes.

**Comment by ghost** [comment]
It seems that the await is unnecessary here? `p.setData()` can be queued and events will happen in the future anyways. And you can check `p.peers` during a state change event.

```
const p = await Presence.start(somehypercore, {
  id: Buffer.from('A', 32)
})
p.setData({something: 'whatever'})
p.on('peer-found', (peer) => peer.data)
p.on('peer-removed', (peer) => peer.id)
for(let peer of p.peers) {
  console.log(peer.id, peer.data, 'connected to', peer.connectedTo)
}
```

**Comment by RangerMauve** [comment]
I like the idea of `registerExtension`, I guess `peer-open` and `peer-remove` events would need to be handled externally somehow?

**Comment by RangerMauve** [comment]
The reason I'm using those events from hypercore is that I need to do some initial bootstrapping when I get a peer, and also listen for when specific peers go offline to notify the rest of the swarm.

**Comment by RangerMauve** [comment]
Hmm, I'm not sure if developing for the stream level extensions instead of the feed level extensions makes sense for me. 🤔 With the stream level, you're basically creating an extension instance per connection and this `.send` only sends to that particular peer, which is great. But with the feed level you need to also track the peers somehow and `.send` ends up being a broadcast to all the peers. 🤷

For my use case I also specifically need to use a feed rather than the stream level since I only want to exchange with peers that know that particular hypercore key and am multiplexing a bunch of hypercores / hyperdrives over the same swarm. 😅

**Comment by RangerMauve** [comment]
Put together [hyper-flood](https://github.com/RangerMauve/hyper-flood). Thanks for the tips for registering the extension @substack. This should work with both hypercore extensions and raw replication protocol extensions. 😁

The presence stuff will just build on top of that.

**Comment by RangerMauve** [comment]
I've got it working over Hypercores now! https://github.com/RangerMauve/hyper-presence/blob/master/index.js

It should be possible to adapt to use with replication streams directly.

You'll want to initialize a hyper-flood instance per connection to start. Might need to get a shared LRU, or see if there's a way to reuse the `extension()` method between connections, not sure how that works for other libraries yet.

This'll be used to broadcast messages through the swarm.

Then you'll want to extend the [Presence class](https://github.com/RangerMauve/hyper-presence/blob/master/Presence.js) and implement a `broadcast` method which will broadcast the message to all your `hyper-flood` instances.

From there you'll want to invoke `this.onAddPeer(id)` and `this.onRemovePeer(id)` whenever you get or lose a connection.

After that, it should be all set. You can invoke `.setData({})` to set some sort of data to share with other peers, and you can use `.online` or `.on('online')` to get the list of peers that are currently online. Also you can get a peer's data with `.getPeerData(id)`. Note that in the hypercore example I'm assuming people's IDs are their NOISE replication public keys. But you can figure out your own as long as all peers can agree on what another peer's ID is.

I'm using this in the Natakanu project to keep track of which archives a given peer in the network is advertising. This is used for automatically discovering content for a group. You could probably stick a person's hypercore key into their data that they use for messages and track when that feed is "active" or "inactive" or something. :P
